core-bundle/src/InsertTag/InsertTagParser.php

```diff
@@ -12,65 +12,629 @@
+            return $tag;
+        }
+
+        if ($tag instanceof ResolvedInsertTag) {
+            return new ParsedInsertTag(
+                $tag->getName(),
+                new ParsedParameters(array_map(static fn ($param) => new ParsedSequence([(string) $param]), $tag->getParameters()->all())),
+                $tag->getFlags(),
+            );
+        }
+
+        throw new \InvalidArgumentException(sprintf('Unsupported insert tag class "%s"', $tag::class));
+    }
+
+    private function resolveParameters(ParsedParameters $parameters): ResolvedParameters
+    {
+        $resolvedParameters = [];
+
+        foreach ($parameters->all() as $parameter) {
+            $value = '';
+
+            foreach ($parameter as $item) {
+                if (\is_string($item)) {
+                    $value .= $item;
+
+                    continue;
+                }
+
+                $value .= $this->renderTag($item)->getValue();
+            }
+
+            $resolvedParameters[] = $value;
+        }
+
+        return new ResolvedParameters($resolvedParameters);
+    }
+
+    /**
+     * @param list<InsertTagResult> $results
+     */
+    private function toChunkedText(array $results): ChunkedText
+    {
+        $chunked = [];
+
+        foreach ($results as $result) {
+            if (OutputType::html === $result->getOutputType()) {
+                $chunked[] = '';
+                $chunked[] = $result->getValue();
+            } else {
+                $chunked[] = $result->getValue();
+                $chunked[] = '';
+            }
+        }
+
+        return new ChunkedText($chunked);
+    }
+
     private function callLegacyClass(string $input, bool $allowEsiTags): ChunkedText
     {
         if (null === $this->insertTags) {
             $this->framework->initialize();
             $this->insertTags = new InsertTags();
         }
 
-        return $this->insertTags->replaceInternal($input, $allowEsiTags);
+        return $this->insertTags->replaceInternal($input, $allowEsiTags, $this);
+    }
+
+    private function handleLegacyTagsHook(ParsedSequence $input, bool $allowEsiTags): ParsedSequence
+    {
+        if (empty($GLOBALS['TL_HOOKS']['replaceInsertTags'])) {
+            return $input;
+        }
+
+        $hasLegacyTags = false;
+
+        foreach ($input as $item) {
+            if (!\is_string($item) && !$this->hasInsertTag($item->getName())) {
+                $hasLegacyTags = true;
+                break;
+            }
+
+            if (\is_string($item) && str_contains($item, '{{')) {
+                $hasLegacyTags = true;
+                break;
+            }
+        }
+
+        if (!$hasLegacyTags) {
+            return $input;
+        }
+
+        $outputs = [];
+
+        foreach ($this->callLegacyClass($input->serialize(), $allowEsiTags) as [$type, $chunk]) {
+            $outputs[] = match ($type) {
+                ChunkedText::TYPE_TEXT => [...$this->doParse($chunk)],
```

Comment:
<img width="685" alt="" src="https://github.com/contao/contao/assets/1192057/f7e2b851-32aa-4943-be0e-f32cb17c6a49">

Should this be `(array) $this->doParse($chunk)` or just `$this->doParse($chunk)`?

Reply:
`[...$x]` is the short version of `iterator_to_array($x)`. I like the short version better but am fine either way.

Reply:
Changed to `iterator_to_array()` in 16cf0c0812acc1769c497db9a583abb7eee9ab8b